pkg/registry/client.go

```diff
@@ -790,6 +791,282 @@ func (c *Client) Tags(ref string) ([]string, error) {
+		sv, err := semver.NewVersion(version)
+
+		// Skip if not valid semver
+		if err != nil {
+			// If we don't have any valid semver yet, use this as fallback
+			if bestSemver == nil && best.tag == "" {
+				best = result
+			}
+			continue
+		}
+
+		// First valid semver
+		if bestSemver == nil {
+			best = result
+			bestSemver = sv
+			continue
+		}
+
+		// Compare with current best
+		// 1. Prefer non-prerelease over prerelease
+		if bestSemver.Prerelease() != "" && sv.Prerelease() == "" {
+			best = result
+			bestSemver = sv
+			continue
+		}
+		if bestSemver.Prerelease() == "" && sv.Prerelease() != "" {
+			continue
+		}
+
+		// 2. Prefer more complete versions (count dots)
+		bestParts := strings.Count(bestSemver.Original(), ".")
+		currentParts := strings.Count(sv.Original(), ".")
+		if currentParts > bestParts {
+			best = result
+			bestSemver = sv
+			continue
+		}
+		if currentParts < bestParts {
+			continue
+		}
+
+		// 3. Prefer higher version when same completeness and release status
+		if sv.GreaterThan(bestSemver) {
+			best = result
+			bestSemver = sv
+		}
+	}
+
+	return best
+}
+
+// Search lists all versions of a chart in an OCI registry and returns metadata
+func (c *Client) Search(ref string, maxVersions int) ([]SearchResult, error) {
+	// Remove oci:// prefix
+	ref = strings.TrimPrefix(ref, OCIScheme+"://")
+
+	ctx := context.Background()
+	repository, err := remote.NewRepository(ref)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create repository for %s: %w", ref, err)
+	}
+	repository.PlainHTTP = c.plainHTTP
+	repository.Client = c.authorizer
+
+	var results []SearchResult
+	digestToMetadata := make(map[string]*chart.Metadata)
+	var digestMutex sync.Mutex
+
+	// Add timeout to prevent hanging on large repositories
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+
+	// First, collect all valid semver tags
+	var validTags []string
+	err = repository.Tags(ctx, "", func(tags []string) error {
+		for _, tag := range tags {
+			// Skip digest tags (sha256-...)
+			if strings.HasPrefix(tag, "sha256-") {
+				continue
+			}
+
+			// Check if it's a valid semver (after converting _ to +)
+			version := strings.ReplaceAll(tag, "_", "+")
+			if _, err := semver.NewVersion(version); err == nil {
+				validTags = append(validTags, tag)
+			}
+		}
+		return nil
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("error listing tags: %w", err)
+	}
+
+	if len(validTags) == 0 {
+		return results, nil
+	}
+
+	// Sort tags by semver (newest first) before processing
+	sort.Slice(validTags, func(i, j int) bool {
+		v1, _ := semver.NewVersion(strings.ReplaceAll(validTags[i], "_", "+"))
+		v2, _ := semver.NewVersion(strings.ReplaceAll(validTags[j], "_", "+"))
+		return v1.GreaterThan(v2)
+	})
+
+	// Process only the specified number of tags to avoid overwhelming large repositories
+	// Users can use --max-versions to increase the limit
+	if maxVersions > 0 && len(validTags) > maxVersions {
```

Comment:
I'm wondering why we are not using [`TagListPageSize`](https://pkg.go.dev/oras.land/oras-go/v2/registry/remote#Repository.TagListPageSize) ?